Mathlib/Tactic/DeriveToExpr.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Lean
+import Mathlib.Init.Align
+
+/-!
+# A `ToExpr` derive handler
+
+This module defines a `ToExpr` derive handler for inductive types. It supports mutually inductive
+types as well.
+
+The `ToExpr` derive handlers support universe level polymorphism. This is implemented using the
+`Lean.ToLevel` class, where `Lean.ToLevel.toLevel.{u}` evaluates to a `Lean.Level` term
+representing `u`. To use `ToExpr` in places where there is universe polymorphism, make sure
+to have a `[ToLevel.{u}]` instance available.
+
+Implementation note: the handler is originally modeled after the `Repr` derive handler.
+-/
+
+namespace Lean
+
+/-- A class to create `Level` expressions that denote particular universe levels in Lean. -/
+class ToLevel.{u} where
+  /-- A `Level` that represents the universe level `u`. -/
+  toLevel : Level
+  /-- The universe itself. This is only here to avoid the "unused universe parameter" error. -/
+  univ : Type u := Sort u
+export ToLevel (toLevel)
+#align reflected_univ Lean.ToLevel
+#align reflected_univ.lvl Lean.ToLevel.toLevel
+
+instance : ToLevel.{0} where
+  toLevel := .zero
+
+instance [ToLevel.{u}] : ToLevel.{u+1} where
+  toLevel := .succ toLevel.{u}
+
+/-
+These instances are dangerous since, for example, the unifier can consider `u =?= max u ?v`.
+Omitting them until it's clear they are needed.
+
+instance (priority := 100) [ToLevel.{u}] [ToLevel.{v}] : ToLevel.{max u v} where
+  toLevel := .max toLevel.{u} toLevel.{v}
+
+instance (priority := 100) [ToLevel.{u}] [ToLevel.{v}] : ToLevel.{imax u v} where
+  toLevel := .imax toLevel.{u} toLevel.{v}
+-/
+
+end Lean
+
+namespace Mathlib.Deriving.ToExpr
+
+open Lean Elab Lean.Parser.Term
+open Meta Command Deriving
+
+def mkToExprHeader (indVal : InductiveVal) : TermElabM Header := do
+  -- The auxiliary functions we produce are `indtype -> Expr`.
+  let header ← mkHeader ``ToExpr 1 indVal
+  return header
+
+/-- Give an expression that is equivalent to `(term|mkAppN $f #[$args,*])` but
+expanded out to use `Expr.app` directly. -/
+def mkAppNExpr (f : Term) (args : Array Term) : TermElabM Term :=
+  args.foldlM (fun a b => `(Expr.app $a $b)) f
```

Comment:
What does equivalent mean here, and why is `(term|mkAppN $f #[$args,*])` not appropriate? I'm afraid I don't have a good mental model of Lean 4 yet, so I'm not actually requesting a change here.

Reply:
It's an optimization. `mkAppN ... #[...]` creates an intermediate array of arguments, and `mkAppNExpr` is unfolding the definition completely. Compare this to the code I posted in an earlier comment:
```
mutual
  private def toExprList.{u} {α} [ToExpr α] [ToLevel.{u}] (x : @List.{u} α) : Expr :=
    match x with
    | @List.nil _ => Expr.app (Expr.const `List.nil [toLevel.{u}]) (toTypeExpr α)
    | @List.cons _ a a' =>
      Expr.app (Expr.app (Expr.app (Expr.const `List.cons [toLevel.{u}]) (toTypeExpr α)) (toExpr a))
        (toExprList a')
end
instance {α} [ToExpr α] [ToLevel.{u}] : ToExpr (@List.{u} α) where
  toExpr := toExprList.{u}
  toTypeExpr := Expr.app (Expr.const `List [toLevel.{u}]) (toTypeExpr α)
```

Reply:
Thanks, I think that's resolved my confusion; we're creating a term that evaluates to an expr, and deciding that we're going to do the expansion up-front when we create the term, rather than at some unspecified future point.

Reply:
Yeah, this derive handler is especially fun in how many different levels of representation it has to wrangle.

Another solution could have been to write a macro that does `mkAppN $f #[$x, $xs,*] => mkAppN (Expr.app $f $x) #[$xs,*]` and `mkAppN $f #[] => $f`, then let that do the unfolding (with some small amount of overhead) rather than us doing it.